packages/core/saved-objects/core-saved-objects-import-export-server-internal/src/export/sort_objects.ts

```diff
@@ -8,33 +8,36 @@
 
 import type { SavedObject } from '@kbn/core-saved-objects-common';
 
+const getId = (object: { type: string; id: string }) => `${object.type}:${object.id}`;
+
 export function sortObjects(savedObjects: SavedObject[]): SavedObject[] {
-  const path = new Set<SavedObject>();
+  const traversed = new Set<string>();
   const sorted = new Set<SavedObject>();
   const objectsByTypeId = new Map(
-    savedObjects.map((object) => [`${object.type}:${object.id}`, object] as [string, SavedObject])
+    savedObjects.map((object) => [getId(object), object] as [string, SavedObject])
   );
 
   function includeObjects(objects: SavedObject[]) {
     for (const object of objects) {
-      if (path.has(object)) {
+      const objectId = getId(object);
+      if (traversed.has(objectId)) {
         continue;
       }
 
-      const refdObjects = object.references
-        .map((ref) => objectsByTypeId.get(`${ref.type}:${ref.id}`))
+      const objectRefs = object.references
+        .map((ref) => objectsByTypeId.get(getId(ref)))
         .filter((ref): ref is SavedObject => !!ref);
 
-      if (refdObjects.length) {
-        path.add(object);
-        includeObjects(refdObjects);
-        path.delete(object);
+      traversed.add(objectId);
+      if (objectRefs.length) {
+        includeObjects(objectRefs);
```

Comment:
This is the significant change of the PR, the rest of the diff in this file is just minor enhancement/cleanup. 

Previously, we were removing the current object from the 'path' when we're done traversing the children, causing the algo to eventually traverse the whole tree again if the same object is encountered in the tree of another node from a higher level.

Once a object has been traversed once, we now keep it in the `path` list (renamed to `traversed` as it better reflect its role now), avoiding the full traversal again later.